sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/rule/JoinReorderingTest.java

```diff
@@ -0,0 +1,462 @@
+  }
+
+  @Test
+  public void testBeamJoinPushThroughJoinRuleRight() throws Exception {
+    RuleSet prepareRules =
+        RuleSets.ofList(
+            SortProjectTransposeRule.INSTANCE,
+            EnumerableRules.ENUMERABLE_JOIN_RULE,
+            EnumerableRules.ENUMERABLE_PROJECT_RULE,
+            EnumerableRules.ENUMERABLE_SORT_RULE,
+            EnumerableRules.ENUMERABLE_TABLE_SCAN_RULE);
+
+    String sqlQuery =
+        "select * from \"tt\".\"medium_table\" as medium_table "
+            + " JOIN \"tt\".\"large_table\" as large_table on large_table.\"medium_key\" = medium_table.\"large_key\" "
+            + " JOIN \"tt\".\"small_table\" as small_table on medium_table.\"small_key\" = small_table.\"medium_key\" ";
+
+    RelNode originalPlan = transform(sqlQuery, prepareRules);
+    RelNode optimizedPlan =
+        transform(
+            sqlQuery,
+            RuleSets.ofList(
+                ImmutableList.<RelOptRule>builder()
+                    .addAll(prepareRules)
+                    .add(BeamJoinPushThroughJoinRule.RIGHT)
+                    .build()));
+
+    assertTopTableInJoins(originalPlan, "small_table");
+    assertTopTableInJoins(optimizedPlan, "large_table");
+  }
+
+  @Test
+  public void testSystemReorderingLargeMediumSmall() {
+    TestTableProvider tableProvider = new TestTableProvider();
+    createThreeTables(tableProvider);
+    BeamSqlEnv env = BeamSqlEnv.withTableProvider(tableProvider);
+
+    // This is Join(Join(large, medium), small) which should be converted to a join that large table
+    // is on the top.
+    BeamRelNode parsedQuery =
+        env.parseQuery(
+            "select * from large_table "
+                + " JOIN medium_table on large_table.medium_key = medium_table.large_key "
+                + " JOIN small_table on medium_table.small_key = small_table.medium_key ");
+    assertTopTableInJoins(parsedQuery, "large_table");
+  }
+
+  @Test
+  public void testSystemReorderingMediumLargeSmall() {
+    TestTableProvider tableProvider = new TestTableProvider();
+    createThreeTables(tableProvider);
+    BeamSqlEnv env = BeamSqlEnv.withTableProvider(tableProvider);
+
+    // This is Join(Join(medium, large), small) which should be converted to a join that large table
+    // is on the top.
+    BeamRelNode parsedQuery =
+        env.parseQuery(
+            "select * from medium_table "
+                + " JOIN large_table on large_table.medium_key = medium_table.large_key "
+                + " JOIN small_table on medium_table.small_key = small_table.medium_key ");
+    assertTopTableInJoins(parsedQuery, "large_table");
+  }
+
+  @Test
+  public void testSystemNotReorderingWithoutRules() {
+    TestTableProvider tableProvider = new TestTableProvider();
+    createThreeTables(tableProvider);
+    ImmutableList.Builder<RelOptRule> ruleSetBuilder = ImmutableList.<RelOptRule>builder();
+    for (RuleSet set : BeamRuleSets.getRuleSets()) {
+      for (RelOptRule rule : set) {
```

Comment:
nit: I think it can be expressed in a declarative way, something like:
```java
set
 .stream()
 .flatMap(set -> set.stream()) //whatever the right way to convert the ruleset to stream of rules
 .filter(rule -> ...)
 .collect(toList());
```
This way it's stateless, has less moving parts, and less nesting

Reply:
Yes, thank you. I just changed it.